base/compiler/typeinfer.jl

```diff
@@ -834,14 +842,10 @@ function typeinf_code(interp::AbstractInterpreter, method::Method, @nospecialize
     mi = specialize_method(method, atypes, sparams)::MethodInstance
     ccall(:jl_typeinf_begin, Cvoid, ())
     result = InferenceResult(mi)
-    frame = InferenceState(result, false, interp)
+    frame = InferenceState(result, :no, interp)
     frame === nothing && return (nothing, Any)
-    if typeinf(interp, frame) && run_optimizer
-        opt_params = OptimizationParams(interp)
-        result.src = src = OptimizationState(frame, opt_params, interp)
-        optimize(interp, src, opt_params, ignorelimited(result.result))
-        frame.src = finish!(interp, result)
-    end
+    run_optimizer && (frame.cached = true)
+    typeinf(interp, frame)
```

Comment:
The previous form here was intended to be an example of running an external optimization pipeline. I think cached=false also has some follow-on effects of deciding how to resolve (and optimize) cycles?

Reply:
Yeah, that's why I made this change. Currently the optimization behavior really changes depending on the cache configuration, and it causes `typeinf_code` to behave slightly different from actual execution, which is very confusing.

For example, https://github.com/JuliaLang/julia/blob/5629856e80c1f50201ca5c775f518593e2b1ad89/test/compiler/inline.jl#L176 was marked as `@test_broken` for long, but I found it has been functional actually
(the problem was that the previous `cached = false` hinders `type_annotate!` to do additional clean up work).

And this change is necessary for the test case added in this PR, because callsite inlining depends on `typeinf_edge`, which actually changes its behavior depending on caller's `cache` configuration.

External consumer can still plugin their own optimization pipeline by overloading `OptimizationState`, so I think this change is okay.